amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+            self.pool.fee_receiver
+        }
+
+        /// Scaled rates are rates multiplied by precision. They are assumed to fit in u128.
+        /// If TOKEN_TARGET_DECIMALS is 18 and RATE_DECIMALS is 12, then rates not exceeding ~340282366 should fit.
+        /// That's because if precision <= 10^18 and rate <= 10^12 * 340282366, then rate * precision < 2^128.
+        fn get_scaled_rates(&self) -> Result<Vec<u128>, MathError> {
+            self.pool
+                .token_rates
+                .iter()
+                .zip(self.pool.precisions.iter())
+                .map(|(rate, &precision)| {
+                    rate.get_rate()
+                        .checked_mul(precision)
+                        .ok_or(MathError::MulOverflow(114))
+                })
+                .collect()
+        }
+
+        fn ensure_onwer(&self) -> Result<(), StablePoolError> {
+            ensure!(
+                self.env().caller() == self.owner,
+                StablePoolError::OnlyOwner
+            );
+            Ok(())
+        }
+
+        fn token_id(&self, token: AccountId) -> Result<usize, StablePoolError> {
+            self.pool
+                .tokens
+                .iter()
+                .position(|&id| id == token)
+                .ok_or(StablePoolError::InvalidTokenId(token))
+        }
+
+        /// Checks if tokens are valid and returns the tokens ids
+        fn check_tokens(
+            &self,
+            token_in: AccountId,
+            token_out: AccountId,
+        ) -> Result<(usize, usize), StablePoolError> {
+            //check token ids
+            let token_in_id = self.token_id(token_in)?;
+            let token_out_id = self.token_id(token_out)?;
+            if token_in_id == token_out_id {
+                return Err(StablePoolError::IdenticalTokenId);
+            }
+            Ok((token_in_id, token_out_id))
+        }
+
+        fn mint_admin_fee(&mut self, fee: u128, token_id: usize) -> Result<(), StablePoolError> {
+            if let Some(fee_to) = self.fee_to() {
+                let admin_fee = self.pool.fees.admin_trade_fee(fee)?;
+                if admin_fee > 0 {
+                    let rates = self.get_scaled_rates()?;
+                    let mut admin_deposit_amounts = vec![0u128; self.pool.tokens.len()];
+                    admin_deposit_amounts[token_id] = admin_fee;
+                    let mut reserves = self.pool.reserves.clone();
+                    reserves[token_id] = reserves[token_id]
+                        .checked_sub(admin_fee)
+                        .ok_or(MathError::SubUnderflow(102))?;
+                    let (admin_fee_lp, _) = math::rated_compute_lp_amount_for_deposit(
+                        &rates,
+                        &admin_deposit_amounts,
+                        &reserves,
+                        self.psp22.total_supply(),
+                        None, // no fees
+                        self.amp_coef(),
+                    )?;
```

Comment:
I'm not sure if it's a problem or not but I noticed the following differences between this implementation and Ref:
1. We subtract `admin_fee` from reserves before calling `rated_compute_lp_amount_for_deposit`, Ref does [not](https://github.com/ref-finance/ref-contracts/blob/main/ref-exchange/src/rated_swap/mod.rs#L660)
2. We don't add `admin_fee` to reserves, Ref [does](https://github.com/ref-finance/ref-contracts/blob/main/ref-exchange/src/rated_swap/mod.rs#L665).